### ISSUE TITLE
MKV Downloads stuck at 99.99%

### ISSUE THREAD

**Issue by OskarEichler** [transmission/transmission#4274]
### What is the issue?

Both in Transmission 4.0.0beta1 as well as Transmission 3 I'm having issues where downloads of MKV files get stuck at 99.99%. It only happens when they are single MKV video files, and when I close Transmission and open it again they suddenly jump to 100% and show as complete. However the actual file still shows as .mkv.part and when removing the .part to have the .mkv file it mostly works but it kind of glitches out quite a bit throughout playback!

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.0 beta1

**Comment by GaryElshaw** [comment]
Whenever i've experienced that it's usually a sign the file has an indeterminant amount of corruption (partially why you might be seeing some visual/audio glitches). Make sure you do a 'verify local data'.

**Comment by OskarEichler** [comment]
I don't think that is the case. The reason the visual/audio is glitched is because the download never finishes. So then I'm the one manually converting the .part into the .mkv which is probably missing some bytes.

**Comment by GaryElshaw** [comment]
Yes, that was my point, all of the data isn't there, causing the glitches. I suspect you have more data missing though. Often, the client will just keep piling more bad blocks on.

**Comment by OskarEichler** [comment]
Hmm potentially - however this only happens with direct MKV files, if they are part of a folder with other files it doesn't happen. It happened consistently for the past 10 or so downloads that all get stuck at 99.99%, even though there are a lot of seeders. Maybe what I will try to do is a use another torrent client for the same link and check if those complete without glitches. Will let you know how that goes.

**Comment by OskarEichler** [comment]
Yep just downloaded the exact same file with Folx and it worked without issues, completed and doesn't have any glitches.

**Comment by GaryElshaw** [comment]
I'm not sure how i can help very much, but happy to. I think this is a known issue.

**Comment by diligiant** [comment]
> I'm not sure how i can help very much, but happy to. I think this is a known issue.
couldn't find it (I'm gonna check again). I would like to contribute to pinpoint and fix the problem (I'm losing a lot of time on this ;) but I don't know where to start.

**Comment by GaryElshaw** [comment]
I raised this as a performance issue recently, and Charles has targeted it for investigation. https://github.com/transmission/transmission/discussions/4154#discussioncomment-4549023

**Comment by GaryElshaw** [comment]
My raison detre for contributions to the project have been to, in my mind, improve the gui of the clients. Throughout that time, in the last year, daily, i'm a little stuck in what and where emphasis should be, or even how how i can contribute as someone who cannot code.

The Transmission project, per se, is constantly evolving, so anything i say today will possibly be irrelevant tomorrow. Today? I wouldn't recommend Transmission 4 as a download client. I think 3 is better. Why?

I think, depending on your gui/ui expectations on your platform, everything else should be in service to connectivity. For downloading, i can regularly see 10% corruption, and often at less than 50% downloaded. I can often see 10x the number of peers with other clients. I can often see not only higher bitrates with other clients, but also more consistent bitrates. This is while pausing a torrent in Transmission and opening the same torrent in another client.

There's work to do.

**Comment by diligiant** [comment]
@GaryElshaw as I'm running it anyway (with 1000+ of simultaneous torrents), is there a debug/telemetry mode I can activate for you so you can observe more, if that's helpful to you.

I patched yesterday with #4178 and that's a lifesaver so far (I know exactly where to stand in terms of "completed" status).

**Comment by GaryElshaw** [comment]
If you're building from source and something has gone awry, the debug info is extremely valuable. https://github.com/transmission/transmission

**Comment by Coeur** [comment]
> I guess you mean this one: #4178 Does this mean that transmission isn't verifying the pieces it receives?

That's what I've been observing, hence why I made this PR. Ideally, it should verify each piece as it receives it, but I don't know how to code that at the moment. If one can deep dive in the code, feel free to improve the behavior.

**Comment by tearfur** [comment]
Is this fixed by #4665?

**Comment by GaryElshaw** [comment]
@OskarEichler @diligiant Are you still seeing this issue in the 4.0.x or 4.1 builds?

**Comment by diligiant** [comment]
I'm running the latest GM and I haven't seen this in a while. I pretty confident as I still patch the GM with code that performs a "verify" when download is complete (I really wished it was rolled in the code base as an option) and I had no 100->99,99 situation.
What a huge relief, **thank you** to everyone btw.

**Comment by GaryElshaw** [comment]
> (I really wished it was rolled in the code base as an option)

I support that idea too, hopefully @coeur will revisit the pr.

**Comment by diligiant** [comment]
Of course as an option. If my memory is correct, it was difficult pre-4 to add an option (with HI).

**Comment by tearfur** [comment]
> Of course as an option. If my memory is correct, it was difficult pre-4 to add an option (with HI).

Feel free to create a feature request issue for more visibility. I'll close this one.
